client/service-container-chain-rpc-provider/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+            (&self.extra_args, &[])
+        }
+    }
+
+    pub fn relaychain_args(&self) -> &[String] {
+        let (relay_chain_args, _) = self.split_extra_args_at_first_dashdash();
+
+        relay_chain_args
+    }
+
+    pub fn orchestrator_chain_args(&self) -> &[String] {
+        let (_, orchestrator_chain_args) = self.split_extra_args_at_first_dashdash();
+
+        orchestrator_chain_args
+    }
+}
+
+pub struct RpcProviderMode<PolkaCli, GRB, RuntimeApi, DVC> {
+    /// General configuration made from container chain arguments
+    pub config: Configuration,
+    pub provider_profile_id: u64,
+
+    /// Run arguments specific for the orchestrator.
+    /// If `None` then the orchestrator is a solochain, it is also the relay chain.
+    pub orchestrator_cli: Option<cumulus_client_cli::RunCmd>,
+
+    /// List of URLs to connect to remote orchestrator nodes. If empty, starts an
+    /// embeded orchestrator node.
+    pub orchestrator_endpoints: Vec<Url>,
+
+    pub collator_options: CollatorOptions,
+
+    /// Run arguments for the relaychain. Will also be the orchestrator chain if
+    /// `orchestrator_cli` is `None`.
+    pub polkadot_cli: PolkaCli,
+
+    /// Run arguments for container chains.
+    pub container_chain_cli: ContainerChainCli,
+
+    /// Generator for the RPC builder, customizable to implement custom RPCs.
+    pub generate_rpc_builder: GRB,
+
+    pub phantom: PhantomData<(RuntimeApi, DVC)>,
+}
+
+impl<PolkaCli, GRB, RuntimeApi, DVC> RpcProviderMode<PolkaCli, GRB, RuntimeApi, DVC>
+where
+    DVC: sc_cli::DefaultConfigurationValues,
+    PolkaCli: sc_cli::CliConfiguration<DVC> + sc_cli::SubstrateCli,
+    RuntimeApi: MinimalContainerRuntimeApi,
+    GRB: GenerateRpcBuilder<RuntimeApi> + 'static,
+{
+    pub async fn run(self) -> sc_cli::Result<TaskManager> {
+        let mut task_manager;
+        let orchestrator_chain_interface: Arc<dyn OrchestratorChainInterface>;
+        let relay_chain_interface: Arc<dyn RelayChainInterface>;
+        let keystore;
+
+        if self.orchestrator_endpoints.is_empty() {
+            // Embeded node
+
+            if let Some(orchestrator_cli) = self.orchestrator_cli {
+                // Parachain orchestrator
+
+                let orchestrator_cli = cli::EmbededOrchestratorCli(orchestrator_cli);
+
+                let tokio_handle = self.config.tokio_handle.clone();
+                let orchestrator_config = SubstrateCli::create_configuration(
+                    &orchestrator_cli,
+                    &orchestrator_cli,
+                    tokio_handle.clone(),
+                )
+                .map_err(|err| format!("Orchestrator chain argument error: {}", err))?;
+
+                let polkadot_config = SubstrateCli::create_configuration(
+                    &self.polkadot_cli,
+                    &self.polkadot_cli,
+                    tokio_handle,
+                )
+                .map_err(|err| format!("Relay chain argument error: {}", err))?;
```

Comment:
Probably worth differentiating the error log here between the different modes we have (parachain, solochain and connection through RPC) like this or with something similar that you consider appropriate.
```suggestion
                .map_err(|err| format!("RpcProvider::parachain_mode: Relay chain argument error: {}", err))?;
```

Reply:
That's good idea. Let's implement both, if possible.